File: DSA/PriyankaSingh_590016827/day20question1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_NAME_LEN 100

// Structure to hold student data
typedef struct 
{
    char name[MAX_NAME_LEN];
    float grade;
} Student;

// Comparison function for qsort (descending by grade)
int compareGrades(const void *a, const void *b) 
{
    Student *s1 = (Student *)a;
    Student *s2 = (Student *)b;

    if (s2->grade > s1->grade)
        return 1;
    else if (s2->grade < s1->grade)
        return -1;
    else
        return 0;
}

int main() 
{
    int n;

    printf("Enter number of students: ");
    scanf("%d", &n);
    getchar(); // Consume newline left by scanf

    Student students[n];

    // Input student data
    for (int i = 0; i < n; i++) 
    {
        printf("Enter name of student %d: ", i + 1);
        fgets(students[i].name, MAX_NAME_LEN, stdin);

        // Removing trailing newline from name
        size_t len = strlen(students[i].name);
        if (len > 0 && students[i].name[len - 1] == '\n') 
        {
            students[i].name[len - 1] = '\0';
        }

        printf("Enter grade of %s: ", students[i].name);
        scanf("%f", &students[i].grade);
        getchar(); // Consuming newline
    }

    // Sort students by grade descending
    qsort(students, n, sizeof(Student), compareGrades);

    printf("\nStudents sorted by grade (highest to lowest):\n");
    for (int i = 0; i < n; i++) {
        printf("%s: %.2f\n", students[i].name, students[i].grade);
    }

    return 0;
}

File: DSA/Vritika_590012956/Day_20_Question_1.py
def sort_students_by_grade(student_records):
    # Sort in  descending order
    return sorted(student_records, key=lambda x: x[1], reverse=True)

n = int(input("Enter the number of students: "))

students = []
for i in range(n):
    name = input(f"Enter name of student {i+1}: ")
    grade = int(input(f"Enter grade of student {i+1}: "))
    students.append((name, grade))

sorted_students = sort_students_by_grade(students)

print("\nStudents sorted by grades (highest first):")
for name, grade in sorted_students:
    print(f"{name}: {grade}")
